### ISSUE TITLE
Texture2D SetData<T> Issues

### ISSUE THREAD

**Issue by Anthonywolfe** [MonoGame/MonoGame#5364]
I was playing with Texture2D's SetData for making slight changes to the texture, 

Here's some code I wrote to copy a Texture to another texture.
```csharp
        public static void Copy(this Texture2D Texture, Rect DstArea, Texture2D Source)
        {
            Color[] SrcChunk = new Color[Source.Width * Source.Height];
            Source.GetData(SrcChunk);

            Rectangle Area = new Rectangle(DstArea.X, DstArea.Y, Source.Width, Source.Height);
            for (int Y = DstArea.Y; Y <= DstArea.Height; Y += Source.Height)
            {
                Area.Y = Y;

                for (int X = DstArea.X; X <= DstArea.Width; X += Source.Width)
                {
                    Area.X = X;
                    
                    Texture.SetData(0, Area, SrcChunk, 0, SrcChunk.Length);
                }
            }
        }
```

This is how I expected it work when a 35x37 image tried to fill in a 200x200 texture
![0](https://c1.staticflickr.com/1/630/31667769805_85e7e9448f_o.png)

This is what I got
![1](https://c1.staticflickr.com/1/445/30826574244_d63961fea6_o.png)

I decided to make sure this wasn't already addressed in the dev branch so I grabbed the latest build (3.6.0.1211)

After changing the code to fall in line with the new requirements of keeping the rectangle in the bounds of the texture

```csharp
        public static void Copy(this Texture2D Texture, Rect DstArea, Texture2D Source)
        {
            Color[] SrcChunk = new Color[Source.Width * Source.Height];
            Source.GetData(SrcChunk);

            Rectangle Area = new Rectangle(DstArea.X, DstArea.Y, Source.Width, Source.Height);
            for (int Y = DstArea.Y; Y <= DstArea.Height; Y += Source.Height)
            {
                Area.Y = Y;
                Area.Height = (Y + Source.Height > Texture.Height) ? Texture.Height - Y : Source.Height;

                for (int X = DstArea.X; X <= DstArea.Width; X += Source.Width)
                {
                    Area.X = X;
                    Area.Width = (X + Source.Width > Texture.Width) ? Texture.Width - X : Source.Width;
                    
                    Texture.SetData(0, Area, SrcChunk, 0, SrcChunk.Length);
                }
            }
        }
```

This is the result I got:
![3](https://c1.staticflickr.com/1/359/30829058904_79faa95967_o.png)

So I'm wondering if the corruption is a bug or not because being able to use a chunk of the texture in edge cases like this would save a expensive GetData call.

**Comment by Jjagg** [comment]
Using a RenderTarget and drawing the first texture repeatedly to the second will be a bunch faster. 
Is this in a DirectX or DesktopGL project?

**Comment by Anthonywolfe** [comment]
DirectX, and yea I'm aware of the RenderTargets being faster. I was just looking for a fallback if for some reason I couldn't use a RenderTarget.

**Comment by willmotil** [comment]
Im pretty sure thats how it worked in xna too i think the loop is vertically centric i always do mine horizontally. The best way to deal with this and avoid it in all cases is to simply create a second color array of the target size assign the pixels from one color array to the other the way you want including not setting out of bounds pixels then drop the entire second larger array to a new texture of the larger size in one shot.

**Comment by Jjagg** [comment]
I'll see if I can find out what's going wrong here and get a fix in with #5121 since that takes care of other Get/SetData issues too.

**Comment by Jjagg** [comment]
The problem is in your implementation. Let's say we have a texture of 4x2 pixels with the following values:

```
1 2 3 4
5 6 7 8
```
When you do GetData you store those pixel values in a linear buffer:
``` 1 2 3 4 5 6 7 8```
Then when you copy that to a 2x2 texture using your method above you take the first 4 values of your buffer and put them in, resulting in:
```
1 2
3 4
```
While you actually wanted
```
1 2
5 6
```
So basically this is happening because Get and SetData work with one dimensional arrays . I hope that clears it up :) You can solve this by either doing another GetData or (better) create another buffer for the edges yourself (this is kind of what wilmotil meant I think). 
This can be closed seeing as it's not a MG bug.
